### ISSUE TITLE
testyield.ml appears to lack robust termination condition

### ISSUE THREAD

**Issue by he32** [ocaml/ocaml#10067]
I've recently been able to make ocaml 4.09.1 build on NetBSD/macppc, but this problem
is also visible on NetBSD/i386 and I'm guessing all the other NetBSD ports where ocaml can be built.

It seems that the `testyield.ml` test lacks a robust termination condition in the
case of a thread scheduler which isn't simply round-robin.

My `make test` run has now been stuck on this single test for a *long* time:

```
  PID USERNAME PRI NICE   SIZE   RES STATE      TIME   WCPU    CPU COMMAND
15121 he        22    4    21M 2392K RUN      276:29 93.90% 93.90% ocamlrun
```

Looking at all the threads with `ps` reveals that it's only 2 of the threads which
accumulate CPU time:

```
granny-smith: {9} ps axsww | egrep "PID|ocamlrun"
 UID   PID  PPID    CPU LID NLWP PRI NI   VSZ   RSS WCHAN    STAT TTY        LTIME COMMAND
1000 15121 23466 124111   6    6  26  4 21704  2392 parked   S-   pts/0  139:08.09 /usr/pkgsrc/lang/ocaml/work/ocaml-4.09.1/runtime/ocamlrun /usr/pkgsrc/lang/ocaml/work/ocaml-4.09.1/testsuite/_ocamltest/tests/lib-systhreads/testyield/ocamlc.byte/testyield.byte 
1000 15121 23466 124111   5    6  29  4 21704  2392 parked   I-   pts/0    0:00.00 /usr/pkgsrc/lang/ocaml/work/ocaml-4.09.1/runtime/ocamlrun /usr/pkgsrc/lang/ocaml/work/ocaml-4.09.1/testsuite/_ocamltest/tests/lib-systhreads/testyield/ocamlc.byte/testyield.byte 
1000 15121 23466 124111   4    6  29  4 21704  2392 parked   I-   pts/0    0:00.00 /usr/pkgsrc/lang/ocaml/work/ocaml-4.09.1/runtime/ocamlrun /usr/pkgsrc/lang/ocaml/work/ocaml-4.09.1/testsuite/_ocamltest/tests/lib-systhreads/testyield/ocamlc.byte/testyield.byte 
1000 15121 23466 124111   3    6  83  4 21704  2392 select   S-   pts/0    0:00.99 /usr/pkgsrc/lang/ocaml/work/ocaml-4.09.1/runtime/ocamlrun /usr/pkgsrc/lang/ocaml/work/ocaml-4.09.1/testsuite/_ocamltest/tests/lib-systhreads/testyield/ocamlc.byte/testyield.byte 
1000 15121 23466 124111   2    6  25  4 21704  2392 -        R-   pts/0  139:08.10 /usr/pkgsrc/lang/ocaml/work/ocaml-4.09.1/runtime/ocamlrun /usr/pkgsrc/lang/ocaml/work/ocaml-4.09.1/testsuite/_ocamltest/tests/lib-systhreads/testyield/ocamlc.byte/testyield.byte 
1000 15121 23466 124111   1    6  29  4 21704  2392 parked   I    pts/0    0:00.02 /usr/pkgsrc/lang/ocaml/work/ocaml-4.09.1/runtime/ocamlrun /usr/pkgsrc/lang/ocaml/work/ocaml-4.09.1/testsuite/_ocamltest/tests/lib-systhreads/testyield/ocamlc.byte/testyield.byte 
1000 21363  7290    269   1    1  85  0  1252    12 pipe_rd  S    pts/4    0:00.02 egrep PID|ocamlrun 
granny-smith: {10} 
```

Looking at the process with `ktruss` reveals it's only threads 2 and 6 which accumulate time, and
constantly switch between being active, plus the very occasional wakeup of thread 3:

```
 15121      6 ocamlrun ___lwp_park60               = 0
 15121      2 ocamlrun ___lwp_park60               = 0
 15121      6 ocamlrun ___lwp_park60               = 0
 15121      2 ocamlrun ___lwp_park60               = 0
 15121      6 ocamlrun ___lwp_park60               = 0
 15121      2 ocamlrun ___lwp_park60               = 0
 15121      6 ocamlrun ___lwp_park60               = 0
 15121      2 ocamlrun ___lwp_park60               = 0
 15121      3 ocamlrun __select50                  = 0
 15121      6 ocamlrun ___lwp_park60               = 0
 15121      6 ocamlrun __sigprocmask14(0x1, 0, 0xfcd33cd0) = 0
 15121      6 ocamlrun __sigprocmask14(0x1, 0xfcd33c80, 0xfcd33c90) = 0
 15121      2 ocamlrun ___lwp_park60               = 0
 15121      6 ocamlrun ___lwp_park60               = 0
 15121      6 ocamlrun __sigprocmask14(0x3, 0xfcd33c90, 0) = 0
 15121      2 ocamlrun ___lwp_park60               = 0
 15121      6 ocamlrun ___lwp_park60               = 0
 15121      2 ocamlrun ___lwp_park60               = 0
 15121      6 ocamlrun ___lwp_park60               = 0
```

So ... I end up having to kill those test cases in order for `make test` to run to completion in finite time.

I'll admit that I don't know ML at all, so I can't really suggest a change to the test case; sorry!

I do however welcome hints, suggestions or fixes...

Best,

Håvard

**Comment by jhjourdan** [comment]
I guess what's happening here is simply that testyield uses a busy-waiting loop for implementing a barrier, which does not work when the pthread_mutex implementation is not fair enough.

Is there a reliable way to implement a barrier from what is provided in the thread module?

**Comment by he32** [comment]
> Is there a reliable way to implement a barrier from what is provided in the thread module?

The `pthread(3)` man page in NetBSD lists

```
           pthread_barrier_destroy(3)    destroy a barrier
           pthread_barrier_init(3)       create a barrier
           pthread_barrier_wait(3)       wait for a barrier
```

The only caveat here is that when calling `pthread_barrier_init` you have to specify how many threads would need to call `pthread_barrier_wait` before they're all released again.

But that was perhaps not what you asked about...  "The thread module" is perhaps an ML thing?  Since I know little to none of ML, I'm not sure I can answer...

**Comment by xavierleroy** [comment]
Yes, a barrier can be implemented using mutexes and condition variables.  IIRC, it's a favorite exam question of @maranget .  

Is the barrier crossed only once, or multiple times?

**Comment by jhjourdan** [comment]
It is crossed only once. I have just implemented that, but it actually does not work for this test, since we not only need a barrier, but we need to make sure all the threads are actually active (and not waiting for the condition variable's mutex) once the barrier is passed.

**Comment by jhjourdan** [comment]
Said otherwise: we need that, when the barrier is passed, then one thread is executing and at least another is waiting on the master lock.

**Comment by jhjourdan** [comment]
One solution would be to use only 2 threads in the test (then busy waiting works well). But then perhaps the test is a bit less relevant.

**Comment by xavierleroy** [comment]
You could have the N threads blocked reading on a pipe, then the main thread writes N bytes to the pipe.  The N threads would then compete directly for the master lock.  

Another solution (to the problem that threads that start early will see a long run without context switches) would be to use statistics.  Each thread builds a histogram of run lengths (row n is how many times a run of n consecutive Thread.yield() without context switch was observed), and an appropriate statistical test is applied to the histograms.

**Comment by jhjourdan** [comment]
@he32 could you please confirm that the test passes if you limit the test to 2 threads?

I.e., could you please replace `let threads = 4` with `let threads = 2` in `testyield.ml` and run the test again?

**Comment by he32** [comment]
I can confirm that if I change `let threads = 4` to `let threads = 2` in `testyield.ml`, the test set no longer gets stuck no this platform.  (Same issue on NetBSD/amd64 btw., so this appears to be a common theme among all the NetBSD platforms.)

**Comment by jhjourdan** [comment]
@ahh, @xavierleroy what would you think of that solution (using only 2 threads for this test)?

The test becomes a little bit less strong, but at least it would work on NetBSD.

**Comment by ahh** [comment]
@jhjourdan I think this test is demonstrating a legitimate issue on NetBSD, actually?  It is pretty scary to me that we entirely starve a thread when all its competitors are aggressively yielding. It is perhaps not one we can practicably fix--I certainly have no knowledge of threading details on NetBSD.  (We could implement our own fair mutex but that sounds miserable.)

I'd actually lean towards disabling the test on netbsd (is that something we can do technically?). Cutting the thread count is an option but I'd definitely want to know if our Linux runtime stopped making progress guarantees with N>2 threads.

**Comment by he32** [comment]
> @jhjourdan I think this test is demonstrating a legitimate issue on NetBSD, actually? It is pretty scary to me that we entirely starve a thread when all its competitors are aggressively yielding. It is perhaps not one we can practicably fix--I certainly have no knowledge of threading details on NetBSD. (We could implement our own fair mutex but that sounds miserable.)

I agree, that doesn't sound either fun or productive.  I'll ask my fellow NetBSD developers what they think of this issue.

> I'd actually lean towards disabling the test on netbsd (is that something we can do technically?). Cutting the thread count is an option but I'd definitely want to know if our Linux runtime stopped making progress guarantees with N>2 threads.

I already have a patch in the pkgsrc package for lang/ocaml which reduces the number of threads to 2, so that the test suite can complete without manual intervention on NetBSD.

**Comment by jhjourdan** [comment]
Note that this is not an issue of the NetBSD scheduler, but rather an "issue" of the implementation of pthread condition variables on NetBSD: recall that the scehduling in OCaml is determined by which thread wins the race of the master lock, not by the system scheduler.

**Comment by he32** [comment]
Hmm, I brought this issue back to NetBSD, and the comment/pushback made was basically `code that makes any assumption about the scheduling after yield is broken`.  I guess what is happening is that the first two threads are stuck in this loop:

```
    while !are_ready < threads do
      Thread.yield ()
    done;
```

preventing the two other threads from becoming active and causing `are_ready` to reach the number of configured threads.

**Comment by ahh** [comment]
Agree, yes: I was lumping in the pthread runtime system + netbsd kernel.

**Comment by he32** [comment]
Sorry I didn't give any feedback earlier.  I'll just note that the earlier comment `code that makes any assumption about the scheduling after yield is broken` means that the test is currently (or was not, I've not checked whether anything has been done to fix this) not coding based on the specification, but coding against a particular implementation and makes non-portable assumptions not prescribed by the standard.
